### PR TITLE
refactor(react): init react package

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

- refactor(react): init react package

## Linear Issue Reference

LOG-1600

Related:
LOG-1627 #204 
LOG-1628 #203 

## Testing

Temporary tests for now.

### CODE REVIEW

File: packages/react/jest.config.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  collectCoverageFrom: ['src/**/*.{js,jsx,ts,tsx}', '!<rootDir>/node_modules/'],
+  coverageReporters: ['text-summary', 'lcov'],
+  moduleFileExtensions: ['ts', 'tsx', 'js'],
+  testEnvironment: 'jsdom',
+  testPathIgnorePatterns: ['lib'],
+  testRegex: '(/__tests__/.*|\\.(test|spec))\\.(ts|tsx|js)$',
+  transform: {
+    '.(ts|tsx)': 'ts-jest',
+  },
+};
```
Comment (gao-sun) [inline packages/react/jest.config.js:11]: 1. should use `.ts`
2. some props are not necessary

File: packages/react/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    "lib": [
+      "dom",
+      "esnext"
+    ],
+    "outDir": "lib",
+    "target": "es5",
+    "types": [
+      "jest"
+    ]
+  },
```
Comment (gao-sun) [inline packages/react/tsconfig.json:13]: this is not needed

**Review by gao-sun — CHANGES_REQUESTED** [review]

File: packages/react/jest.config.js (hunk 1 above)
Comment (IceHe) [reply packages/react/jest.config.js:11]: image
If `js/react` provide `Provider`, it should contain `.tsx` files?

File: packages/react/package.json [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    "@types/react": "^17.0.39",
+    "@types/react-dom": "^17.0.11",
+    "@types/react-router-dom": "^5.3.3",
+    "codecov": "^3.8.3",
+    "eslint": "^8.9.0",
+    "jest": "^27.0.6",
+    "lint-staged": "^12.3.4",
+    "parcel": "^2.3.2",
+    "prettier": "^2.5.1",
+    "stylelint": "^13.13.1",
```
Comment (demonzoo) [inline packages/react/package.json:46]: I don' think we need to lint styles in react SDK. Consider removing it

**Review by demonzoo — APPROVED** [review]
Overall looks good!

File: packages/react/jest.config.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+  coverageReporters: ['text-summary', 'lcov'],
```
Comment (simeng-li) [inline packages/react/jest.config.js:2]: `collectCoverageFrom` config is required I believe. If I remember correctly, if it is not provided, if a target file is not referenced by any test file, it won't be counted into the coverage collection.

**Review by simeng-li — CHANGES_REQUESTED** [review]

File: packages/react/jest.config.js (hunk 4 above)
Comment (IceHe) [reply packages/react/jest.config.js:2]: Will use `coveragePathIgnorePatterns: ['/lib/', '/node_modules/']` instead.

File: packages/react/package.json (hunk 3 above)
Comment (IceHe) [reply packages/react/package.json]: Removed.

(Used to fix missing peer in eslint-config-react. Seems to be unnecessary to fix it here.) 

image

File: packages/react/jest.config.js (hunk 4 above)
Comment (IceHe) [reply packages/react/jest.config.js]: Updated. PTAL.

File: packages/react/tsconfig.json (hunk 2 above)
Comment (IceHe) [reply packages/react/tsconfig.json]: Simplified. PTAL.

File: packages/react/jest.config.js (hunk 1 above)
Comment (IceHe) [reply packages/react/jest.config.js]: Updated. PTAL.

File: packages/react/package.json [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    "package": "parcel build",
+    "test": "jest",
+    "test:coverage": "jest --silent --coverage",
+    "test:report": "codecov -F react",
+    "prepublish": "pnpm build && pnpm package",
+    "report": "WITH_REPORT=true pnpm package"
+  },
+  "dependencies": {
+    "@logto/browser": "^0.1.0",
+    "react": "^17.0.2",
```
Comment (wangsijie) [inline packages/react/package.json:24]: Do we need to move `react` to peerDependencies? @logto-io/eng

Comment (demonzoo) [reply packages/react/package.json:24]: Yes, I think we do, as this is `react SDK`, and it is supposed to provide hooks and router components. @wangsijie

Comment (gao-sun) [reply packages/react/package.json:24]: agreed

File: packages/react/package.json [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  "exports": "./lib/index.js",
+  "typings": "lib/index.d.ts",
+  "license": "MPL-2.0",
+  "scripts": {
+    "dev:tsc": "tsc -p tsconfig.build.json -w --preserveWatchOutput",
+    "preinstall": "npx only-allow pnpm",
+    "precommit": "lint-staged",
+    "build": "tsc -p tsconfig.build.json",
+    "lint": "eslint --ext .ts --ext .tsx src",
+    "package": "parcel build",
```
Comment (gao-sun) [inline packages/react/package.json:15]: why do we need parcel here if we just use `tsc` for compiling?

File: packages/react/package.json [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    "test:coverage": "jest --silent --coverage",
+    "test:report": "codecov -F react",
+    "prepublish": "pnpm build && pnpm package",
+    "report": "WITH_REPORT=true pnpm package"
+  },
+  "dependencies": {
+    "@logto/browser": "^0.1.0",
+    "react": "^17.0.2",
+    "react-dom": "^17.0.2",
+    "react-router-dom": "^6.2.1"
```
Comment (gao-sun) [inline packages/react/package.json:26]: also `react-router-dom` should NOT present in dependency for now? can't see the usage

File: packages/react/package.json [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  },
+  "devDependencies": {
+    "@jest/types": "^27.5.1",
+    "@parcel/core": "^2.3.2",
+    "@silverhand/eslint-config": "^0.9.1",
+    "@silverhand/eslint-config-react": "^0.9.1",
+    "@silverhand/ts-config": "^0.9.1",
+    "@silverhand/ts-config-react": "^0.9.1",
+    "@testing-library/react": "^12.1.3",
+    "@testing-library/react-hooks": "^7.0.2",
```
Comment (gao-sun) [inline packages/react/package.json:36]: testing library looks not needed for now?

File: packages/react/package.json [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    "test": "jest",
+    "test:coverage": "jest --silent --coverage",
+    "test:report": "codecov -F react",
+    "prepublish": "pnpm build && pnpm package",
+    "report": "WITH_REPORT=true pnpm package"
+  },
+  "dependencies": {
+    "@logto/browser": "^0.1.0",
+    "react": "^17.0.2",
+    "react-dom": "^17.0.2",
```
Comment (gao-sun) [inline packages/react/package.json:25]: i'm not sure if we need `react-dom` in our dependency list if we just ship hooks / components

Comment (IceHe) [reply packages/react/package.json:25]: Removed for now. PTAL

File: packages/react/package.json (hunk 5 above)
Comment (IceHe) [reply packages/react/package.json]: Moved to peerDependencies. PTAL

File: packages/react/package.json (hunk 6 above)
Comment (IceHe) [reply packages/react/package.json]: Removed parcel for now.

File: packages/react/package.json (hunk 7 above)
Comment (IceHe) [reply packages/react/package.json]: Removed for now. PTAL

File: packages/react/package.json (hunk 8 above)
Comment (IceHe) [reply packages/react/package.json]: Removed for now. PTAL

File: packages/react/webpack.config.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+const { BundleAnalyzerPlugin } = require('webpack-bundle-analyzer');
+
+module.exports = () => {
+  const { WITH_REPORT } = process.env;
+
+  return {
+    mode: 'production',
+    entry: './src/index.ts',
+    output: {
+      path: path.resolve(__dirname, 'dist'),
```
Comment (gao-sun) [inline packages/react/webpack.config.js:12]: why do we have 2 different output path? (here `dist` while tsconfig holds `lib`)

File: packages/react/webpack.config.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+      filename: 'logto.react.min.js',
+      library: {
+        type: 'umd',
+        name: 'logto',
```
Comment (gao-sun) [inline packages/react/webpack.config.js:16]: should be `logto.react` or `logto-react`?

File: packages/react/webpack.config.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+      },
+    },
+    module: {
+      rules: [
+        {
+          exclude: /node_modules/,
+          use: 'ts-loader',
```
Comment (gao-sun) [inline packages/react/webpack.config.js:23]: do we still need `ts-loader` if we already use `tsc` to compile files to js?

File: packages/react/webpack.config.js (hunk 11 above)
Comment (IceHe) [reply packages/react/webpack.config.js:16]: Follow the [discussion](https://silverhand-io.slack.com/archives/C02A8G4HVAM/p1644208989260219) on Slack

image

File: packages/react/webpack.config.js (hunk 10 above)
Comment (IceHe) [reply packages/react/webpack.config.js]: Removed `Webpack`. Do not support `package` for now and add it when needed.

File: packages/react/webpack.config.js (hunk 12 above)
Comment (IceHe) [reply packages/react/webpack.config.js]: Removed `Webpack`. Do not support `package` for now and add it when needed.

File: packages/react/webpack.config.js (hunk 11 above)
Comment (IceHe) [reply packages/react/webpack.config.js]: Removed `Webpack`. Do not support `package` for now and add it when needed.

File: packages/react/package.json [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    "ts-jest": "^27.0.4",
+    "typescript": "^4.5.5"
+  },
+  "peerDependencies": {
+    "react": ">=16.8.0"
+  },
+  "eslintConfig": {
+    "extends": "@silverhand"
+  },
+  "prettier": "@silverhand/eslint-config/.prettierrc"
```
Comment (gao-sun) [inline packages/react/package.json:45]: should use react config here?

Comment (IceHe) [reply packages/react/package.json:45]: Updated. PTAL

File: packages/react/jest.config.ts [hunk 14]
```diff
@@ -0,0 +1,13 @@
+import type { Config } from '@jest/types';
+
+const config: Config.InitialOptions = {
+  preset: 'ts-jest',
+  collectCoverageFrom: ['src/**/*.{ts|tsx}'],
+  coverageReporters: ['lcov', 'text-summary'],
+  testEnvironment: 'jsdom',
+  transform: {
```
Comment (IceHe) [inline packages/react/jest.config.ts:8]: block comment

File: packages/react/jest.config.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+    '\\.(ts|js)x?$': 'ts-jest',
+  },
+};
```
Comment (IceHe) [inline packages/react/jest.config.ts:11]: single line comment

File: packages/react/package.json (hunk 13 above)
Comment (IceHe) [reply packages/react/package.json]: Updated. PTAL.

**Review by gao-sun — APPROVED** [review]

**Review by simeng-li — APPROVED** [review]

**Pull request merged**